<launch>
  <!-- 引入其他的launch文件 -->
  <!-- 同时实现amcl和fht_map -->
  <!-- 每次实验修改一下初始位置 -->
  <env name="GAZEBO_MODEL_PATH" value="$(find turtlebot3sim)/models/"/>
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="x_pos" default="-3.0"/>
  <arg name="y_pos" default="1.0"/>
  <arg name="z_pos" default="0.0"/>
  

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find turtlebot3sim)/worlds/large_indoor.world"/>
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <!-- 每次实验修改一下初始位置 -->
  <arg name="origin_x" value="6.0"/>
  <arg name="origin_y" value="9"/>
  <arg name="origin_yaw" value="0.0"/>


  <group ns="robot1">
      <include file="$(find turtlebot3sim)/launch/includes/robot.launch.xml">
          <arg name="robot_name" value="robot1"/>
          <arg name="model" value="$(arg model)"/>
          <arg name="init_pose" value="-x $(arg origin_x) -y $(arg origin_y) -z 0.12 -Y $(arg origin_yaw)"/>
          <!-- <arg name="init_pose" value="-x 7 -y 8.0 -z 0.0 -Y 1.57"/> -->
      </include>
  </group>

  <include file="$(find turtlebot3sim)/launch/robots_slam/single_robot_origin.launch" />
  
  
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find fht_map)/rviz/topo_reloca.rviz" />


  <node name="map_building" pkg="fht_map" type="topo_relocalization.py" output="screen">
    <param name="network" value="retrievalSfM120k-vgg16-gem" />
    <param name="platform" value="cuda" />
    <param name="robot_name" value="robot1"/>
    <param name="robot_num" value="1"/>
  </node>

  <!-- 运行地图服务器，并且加载设置的地图-->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find fht_map)/data/map_data/large_indoor/map.yaml">
        <param name="frame_id" value="reloca_map"/>
    </node>

    <node pkg="tf" type="static_transform_publisher" name="tf_publisher1" args="0 0 0 0 0 0 /robot1/odom /robot1/odom_amcl 100" />

    <node pkg="amcl" type="amcl" name="amcl" output="screen">
        <!-- Publish scans from best pose at a max of 10 Hz -->
        <param name="odom_model_type" value="diff"/><!-- 里程计模式为差分 -->
        <param name="odom_alpha5" value="0.1"/>
        <param name="transform_tolerance" value="0.2" />
        <param name="gui_publish_rate" value="10.0"/>
        <param name="laser_max_beams" value="30"/>
        <param name="min_particles" value="500"/>
        <param name="max_particles" value="5000"/>
        <param name="kld_err" value="0.05"/>
        <param name="kld_z" value="0.99"/>
        <param name="odom_alpha1" value="0.2"/>
        <param name="odom_alpha2" value="0.2"/>
        <!-- translation std dev, m -->
        <param name="odom_alpha3" value="0.8"/>
        <param name="odom_alpha4" value="0.2"/>
        <param name="laser_z_hit" value="0.5"/>
        <param name="laser_z_short" value="0.05"/>
        <param name="laser_z_max" value="0.05"/>
        <param name="laser_z_rand" value="0.5"/>
        <param name="laser_sigma_hit" value="0.2"/>
        <param name="laser_lambda_short" value="0.1"/>
        <param name="laser_lambda_short" value="0.1"/>
        <param name="laser_model_type" value="likelihood_field"/>
        <!-- <param name="laser_model_type" value="beam"/> -->
        <param name="laser_likelihood_max_dist" value="2.0"/>
        <param name="update_min_d" value="0.2"/>
        <param name="update_min_a" value="0.5"/>
 
        <!--设置坐标系：odom、map 和 机器人基坐标系-->
        <param name="odom_frame_id" value="/robot1/odom_amcl"/><!-- 里程计坐标系 -->
        <param name="base_frame_id" value="/robot1/base_footprint"/><!-- 添加机器人基坐标系 -->
        <param name="global_frame_id" value="reloca_map"/><!-- 添加地图坐标系 -->
        <remap from="/scan" to="/robot1/scan"/>
 
        <param name="resample_interval" value="1"/>
        <param name="transform_tolerance" value="0.1"/>
        <param name="recovery_alpha_slow" value="0.0"/>
        <param name="recovery_alpha_fast" value="0.0"/>
    </node>


    <node name="rviz2" pkg="rviz" type="rviz" args="-d $(find fht_map)/rviz/acml.rviz">
    </node>

    <node name="experiment2" pkg="fht_map" type="exp_reloca_amcl.py" output="screen">
        <param name="origin_x" value="$(arg origin_x)"/>
        <param name="origin_y" value="$(arg origin_y)"/>
        <param name="sim_env" value="large_indoor"/>
        <param name="origin_yaw" value="$(arg origin_yaw)"/>
        <param name="robot_num" value="1"/>
        <param name="reloca_method" value="amcl"/>
    </node>

  <node pkg="rosservice" type="rosservice" name="global_localization" args="call /global_localization" />
</launch>